### PR TITLE
feat: initial support for dependent resources

### PR DESCRIPTION
**Pull request by metacosm**
This PR provides annotation-based configuration of dependent resources, which the controller then automatically handles, taking care of propagating creation/update of said dependent resources depending on how they are configured. The functional scope is still pretty limited and more changes will be needed but this sets up the architecture for more changes.

In particular, handling of readiness / status reporting of dependent resources will also be tackled at some point.

### CODE REVIEW

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/DefaultContext.java [hunk 1]
```diff
@@ -2,16 +2,39 @@
...
+import io.javaoperatorsdk.operator.processing.event.source.EventSourceRegistry;
+
+public class DefaultContext<P extends HasMetadata> implements Context<P> {
 
   private final RetryInfo retryInfo;
+  private final Controller<P> controller;
+  private final P primaryResource;
 
-  public DefaultContext(RetryInfo retryInfo) {
+  public DefaultContext(RetryInfo retryInfo, Controller<P> controller, P primaryResource) {
```
Comment (csviri) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/DefaultContext.java:16]: Should this go to primary context? Cannot be an implementation of `Reconciler` that prodives these? 
To have the current default  APIs unchanged? Wouldn't that be possible?

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/DefaultContext.java [hunk 2]
```diff
@@ -2,16 +2,38 @@
...
     return Optional.ofNullable(retryInfo);
   }
+
+  @Override
+  public EventSourceRegistry<P> getEventSourceRegistry() {
+    return controller.getEventSourceRegistry();
+  }
+
+  @Override
+  public <T extends HasMetadata> T getSecondaryResource(Class<T> expectedType,
```
Comment (csviri) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/DefaultContext.java:33]: What is difference between dependent and secondary resource?

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/EventSourceInitializer.java [hunk 3]
```diff
@@ -12,6 +13,6 @@
    * @param eventSourceRegistry the {@link EventSourceRegistry} where event sources can be
    *        registered.
    */
-  void prepareEventSources(EventSourceRegistry<T> eventSourceRegistry);
+  void prepareEventSources(EventSourceRegistry<T> eventSourceRegistry, Cloner cloner);
```
Comment (csviri) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/EventSourceInitializer.java:16]: Why is the cloner needed here? Is not generic for the whole operator?

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/DependentResource.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
+package io.javaoperatorsdk.operator.api.reconciler.dependent;
+
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.javaoperatorsdk.operator.api.config.dependent.DefaultDependentResourceConfiguration;
+import io.javaoperatorsdk.operator.api.config.dependent.DependentResourceConfiguration;
+import io.javaoperatorsdk.operator.processing.event.ResourceID;
+import io.javaoperatorsdk.operator.processing.event.source.ResourceCache;
+import io.javaoperatorsdk.operator.processing.event.source.ResourceEventSource;
+
+public class DependentResource<R extends HasMetadata, P extends HasMetadata> {
```
Comment (csviri) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/DependentResource.java:10]: P is the parent? Or what it is used for?

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/Controller.java [hunk 5]
```diff
@@ -1,43 +1,81 @@
...
+  private void waitUntilStarted() {
+    if (!started.get()) {
+      AtomicInteger count = new AtomicInteger(0);
+      final var waitTime = 50;
+      while (!started.get()) {
+        try {
+          count.getAndIncrement();
+          Thread.sleep(waitTime);
+        } catch (InterruptedException e) {
+          e.printStackTrace();
```
Comment (csviri) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/Controller.java:68]: Should be a log message, or even fail probably

**Comment by metacosm** [comment]
Closing this in favor of #785

**Pull request closed without merging**